bin/diagnostics.bash

```diff
@@ -7,32 +7,38 @@ here="$(dirname "$(readlink -f "$0")")"
 source "${here}/common.bash"
 
 usage() {
-    log_info "usage: ck8s diagnostics <sc|wc> [--namespace namespace] [--include-config]"
+    log_info "usage: ck8s diagnostics <sc|wc> [--namespace namespace] [--include-config] [--with-metrics-since date] [--query-metric metric]"
     log_info
     log_info "Collects diagnostics from the current environment set by CK8S_CONFIG_PATH and"
     log_info "store them in a file in the CK8S_CONFIG_PATH directory encrypted with SOPS using"
     log_info "any GPG keys listed in CK8S_CONFIG_PATH/.sops.yaml or set by CK8S_PGP_FP"
     log_info ""
-    log_info "\t-h, --help\t\tdisplay help for this command and exit"
-    log_info "\t-n, --namespace\t\trun diagnostics for specified namespace only"
-    log_info "\t    --include-config\tinclude config yaml files found in CK8S_CONFIG_PATH"
+    log_info "\t-h, --help\t\t\tdisplay help for this command and exit"
+    log_info "\t-n, --namespace\t\t\ttrun diagnostics for specified namespace only"
+    log_info "\t    --include-config\t\tinclude config yaml files found in CK8S_CONFIG_PATH"
+    log_info "\t    --with-metrics-since\tquery a predefined set of metrics since the specified date"
+    log_info "\t    --query-metric\t\tquery any arbitrary metric"
     exit 1
 }
 
 include_config=false
-namespace=""
 cluster="${1}"
+sub_command=""
+command_arg=""
+
 shift
 
 while [ "${#}" -gt 0 ] ; do
     case "${1}" in
-        -n | --namespace)
-            namespace="${2:-}"
-            shift
-            ;;
         --include-config)
             include_config=true
             ;;
+        namespace|with-metrics-since|query-metric)
+            [[ ${#} -lt 2 || "${2}" = -* || -n "$sub_command" ]] && usage
```

Comment:
How is this check supposed to work?
Shouldn't it be `[[ check that all is alright ]] || usage`, where `usage` terminates the script?

Reply:
Its just the opposite, works the same. Do you want me to flip it?
`[[ is anything wrong ]] && usage`

Reply:
Did you validate this?
Because with the bash options we usually set, if a command returns `1` then the script should fail.
And if the check `[[ is anything wrong fail ]]` fails (so when all is good) then it should short circuit.

Reply:
I had to dig into some documentation about this and I am wrong, it does work, but at the same time I don't understand why it manages to work.

> An AND list has the form `command1 && command2`, _command2_ is executed if, and only if, _command1_ returns an exit status of zero (success).
> ...
> The return status of AND and OR lists is the exit status of the last command executed in the list.

So that happens, however as I tested a similar construct bash does indeed not treat that as a failure that will cause the script to terminate.
Even though the status after the check is `1`.

I would prefer if we used constructs that behave according to what should be right, rather than this exception.

Reply:
I did some more digging, apparently `errexit` explicitly allows this:
> The shell does not exit if the command that fails is part of the command list immediately following a while or until keyword, part of the test in an if statement, part of any command executed in a && or || list except the command following the final && or ||, any command in a pipeline but the last, or if the command’s return status is being inverted with !.

But I keep my stance on it.

Reply:
I've tested this and it works, and I'm not quite sure I understand what you mean.
The check currently makes sense to me in my head. We have some checks that can't be true, if any is true it short circuits to usage, usage prints usage and exits. If all checks are fine (i.e false) it does not run usage and continues on.
That is how an if statement makes sense to me, e.g.
```
if (somethings wrong) then {
    print error
    exit
}

rest of function
```


But I see we use `|| usage` everywhere else so I'll try to flip it just for sake of consistency.

Reply:
Yes, but here you are leaving out the actual if statement, and are instead running two commands chained by `&&`, which makes it different.
If you'd rewrite it to use an if statement then that would be fine with me as an alternative.

With `errexit` which we set, if a command fails the script fails.
However adding `&&` inhibits this behaviour.
And this is what's not making sense to me.

So normally we expect commands to cause the script to exit if the commands of a line fails.
However the happy path with `[[ error check ]] && usage` is that the line _should_ fail and not exit the script.
And I want to avoid constructs that use these exceptions to make it clearer on how they work.
That is the difference with `[[ ok check ]] || usage`, as that line succeeds if everything is ok.

Not sure if that makes it clearer or just more confusing. :sweat_smile:

Reply:
Ah I see, I understand. But does `||` not also inhibit this behaviour but just reversed? So those line also does not really make sense to me then with `errexit` set. The [Bash reference](https://www.gnu.org/software/bash/manual/bash.html#index-trap) seems to indicate that at least.
We explicitly exit in usage in both cases so we are not really relying on `errexit` anyway, or am I misunderstanding?

I'll change no matter what just to be the same everywhere just trying to make things clear 😅.

Reply:
Yeah I guess I understand. In my head just looking at the lines I still think my original way makes more sense as I see it as the if case but I guess that ignores bash being bash.

Reply:
> Yes, but I think the overall logic makes more sense there as we then have `true || ?`. The overall expression is true, so it make sense to continue.
> 
> Compare it to `false && ?` where the overall logic says `false`.

`true || ?` does not evaluate `?` in this case. It's the other way around:
```bash
$ true && echo foo
foo
$ false && echo foo
$ true || echo foo
$ false || echo foo
foo
```

Reply:
To me it makes perfect sense to use `[ ... ] || usage` => "check if command line is correct, else show usage".

Reply:
> > Yes, but I think the overall logic makes more sense there as we then have `true || ?`. The overall expression is true, so it make sense to continue.
> > Compare it to `false && ?` where the overall logic says `false`.
> 
> `true || ?` does not evaluate `?` in this case. It's the other way around:
> 
> ```shell
> $ true && echo foo
> foo
> $ false && echo foo
> $ true || echo foo
> $ false || echo foo
> foo
> ```

My bad, should have read your message with the entire thread in context. I get what you mean with `true || ?` now. :smile: